jsrc/conversions.cpp

```diff
@@ -0,0 +1,637 @@
+/* Copyright 1990-2010, Jsoftware Inc.  All rights reserved.               */
+/* Licensed use only. Any other use is in violation of copyright.          */
+/*                                                                         */
+/* Conversions Amongst Internal Types                                      */
+
+#include <algorithm>
+#include <limits>
+#include <numeric>
+#include <optional>
+
+#include "array.hpp"
+extern "C" {
+#include "verbs/vcomp.h"
+}
+
+#define CVCASE(a, b) (((a) << 3) + (b))  // The main cases fit in low 8 bits of mask
+
+// fuzzy_equal() is used in bcvt, where FUZZ may be set to 0 to ensure only exact values are demoted to lower precision
+// used when v is known to be exact integer. It's close enough, maybe ULP too small on the high end
+[[nodiscard]] static auto
+fuzzy_equal(double u, double v, double fuzz) -> bool {
+    return std::abs(u - v) <= fuzz * std::abs(v);
+}
+
+template <typename T, typename V>
+[[nodiscard]] constexpr auto
+in_range(V value) -> bool {
+    if constexpr (std::is_same_v<bool, V>) {
+        (void)value;
+        return true;
+    } else {
+        return std::numeric_limits<T>::lowest() <= value && value <= std::numeric_limits<T>::max();
+    }
+}
+
+template <typename T, typename V>
+[[nodiscard]] constexpr auto
+in_range() -> bool {
+    return in_range<T>(std::numeric_limits<V>::lowest()) && in_range<T>(std::numeric_limits<V>::max());
+}
+
+template <typename T>
+struct is_optional : std::false_type {};
+
+template <typename T>
+struct is_optional<std::optional<T>> : std::true_type {};
+
+template <typename T>
+[[nodiscard]] auto
+value_if(bool cond, T value) -> std::optional<T> {
+    return cond ? std::optional(value) : std::nullopt;
+}
+
+template <typename From, typename To, typename Transform>
+[[nodiscard]] auto
+convert(J jt, array w, void *yv, Transform t) -> bool {
+    auto *v = pointer_to_values<From>(w);
+    auto *result = static_cast<To *>(yv);
+    if constexpr (is_optional<decltype(t(*v))>::value) {
+        for (int64_t i = 0; i < AN(w); ++i) {
+            auto opt = t(v[i]);
+            if (!opt) return false;
+            result[i] = opt.value();
+        }
+    } else {
+        std::transform(v, v + AN(w), result, t);
+    }
+    return true;
+}
+
+template <typename From, typename To>
+[[nodiscard]] auto
+convert(J jt, array w, void *yv) -> bool {
+    if constexpr (std::is_same_v<To, Z>) {
+        return convert<From, To>(jt, w, yv, [](auto v) { return To{.re = static_cast<D>(v), .im = {}}; });
+    } else if constexpr (!in_range<To, From>()) {
+        return convert<From, To>(jt, w, yv, [](auto v) { return value_if(in_range<To>(v), v); });
+    } else {
+        (void)jt;
+        auto *v = pointer_to_values<From>(w);
+        std::copy(v, v + AN(w), static_cast<To *>(yv));
+        return true;
+    }
+}
+
+template <>
+[[nodiscard]] auto
+convert<D, bool>(J jt, array w, void *yv, D fuzz) -> bool {
```

Comment:
Let's try to avoid the D macro
```suggestion
convert<D, bool>(J jt, array w, void *yv, double fuzz) -> bool {
```

Reply:
Maybe same goes for all the others

Reply:
Did `D` -> `double` and `I` -> `int64_t` (and one last `B` -> `bool`)

most others are j specific structs

ps. typedefs not macros

Reply:
The different width integer types (1, 2, 3, 4 bytes; UC, US...) could possibly be replaced, but seeing the compiler, I'd postpone that to whenever those typedefs are refactored.